### PR TITLE
System.Composition Activator parameter initialization race condition

### PR DESCRIPTION
**Pull request by jswolf19**
Fix for issue #6857 where the `ParameterInfo`s from multiple calls to `ConstructorInfo.GetParameters()` are incorrectly assumed to be reference equal. @AlexGhiondea

### CODE REVIEW

**Comment by danmosemsft** [comment]
@AlexGhiondea can you review and merge if appropriate

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs [hunk 1]
```diff
@@ -226,5 +226,44 @@ public bool TryCloseGenericPart(Type[] typeArguments, out DiscoveredPart closed)
...
+                {
+                    return false;
+                }
+
+                if (x.Position != y.Position)
+                {
+                    return false;
+                }
+
+                if (!ReferenceEquals(x.Member, y.Member))
```
Comment (jswolf19) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: I particularly want to know if this should be done differently (by using Equals, or omitted completely)...

**Comment by jswolf19** [comment]
@hughbe I think I screwed up on the rebase... I'm not too familiar with git. Is there something I can do to fix this so that the files changed and all isn't screwed up?

**Comment by hughbe** [comment]
I've done that too many times - Try checking out your branch, then typing `git rebase dotnet/maste` the force pushing with `git push -f` :)

**Comment by jswolf19** [comment]
@hughbe Thanks. Much nicer now ^_^

**Comment by stephentoub** [comment]
cc: @AtsushiKan

**Comment by karelz** [comment]
What is the next step for this PR? @AlexGhiondea can you please push it forward? Thanks!

**Comment by karelz** [comment]
@AlexGhiondea ping?

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs (hunk 1 above)
Comment (AlexGhiondea) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: I think you can use x.Member == y.Member here. MemberInfo overrides the `==` operator and does a few more checks.

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs [hunk 2]
```diff
@@ -226,5 +226,44 @@ public bool TryCloseGenericPart(Type[] typeArguments, out DiscoveredPart closed)
...
+        // between ParameterInfo objects from the same constructor reference,
+        // thus only the position needs to be compared.
+        // Equals checks the member reference equality in case usage changes.
+        private sealed class ParameterInfoComparer : IEqualityComparer<ParameterInfo>
+        {
+            public static readonly ParameterInfoComparer Instance = new ParameterInfoComparer();
+
+            public int GetHashCode(ParameterInfo obj)
+            {
+                return obj.Position.GetHashCode();
```
Comment (AlexGhiondea) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: My preference would be to not rely on the current usage of this when designing the comparer.

Comment (jswolf19) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: If you mean you'd prefer a general `EqualityComparer` that would provide a better hash if the `ParameterInfo`s aren't from the same member, then I can add in `Member.GetHashCode()`. I think I remember seeing something about a common class somewhere for combining hashes. Do you happen to know what I'm referring to, or should I just go with an xor?

Comment (AlexGhiondea) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: Maybe this one? 
https://github.com/dotnet/corefx/blob/8f824ba00f5656436f7063da49fb0ac94eff8566/src/Common/src/System/Numerics/Hashing/HashHelpers.cs

**Comment by danmosemsft** [comment]
@AlexGhiondea looks like feedback addressed

**Comment by karelz** [comment]
@AlexGhiondea can you please merge it?

**Comment by jswolf19** [comment]
@AlexGhiondea no problem. Looking forward to picking up another in the future.

**Pull request merged**